### PR TITLE
CI/CD

### PR DESCRIPTION
**Pull request by kristojorg**
Issue number: #147 

## **This PR does the following:**
- Adds a CI workflow to lint, typecheck and test code on all pushes and pull requests.
- Adds a release workflow to tag, release, publish to npm, and upload to github pages

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ public
 .DS_Store
 lib
 dist
+storybook-static
```
Comment (helenvholmes) [inline .gitignore:7]: Since we're adding this to the `.gitignore`, can you also remove the last section about `storybook-static` from the README?

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
+# This workflow will tag, release, publish to npm, and upload to gh pages.
+name: Production Release
+
+# workflow will only run when a pull request to the release branch is closed.
+on:
+    push:
+        branches: [release]
```
Comment (helenvholmes) [inline .github/workflows/release.yml:7]: Does this mean that instead of a `master`, we have a branch that matches the name of the release? Is that the idea?

**Review by helenvholmes — CHANGES_REQUESTED** [review]
I have a few elementary questions that I think could be answered with some documentation. @kristojorg, as part of this PR, can you also update [this wiki page](https://github.com/NYPL/nypl-design-system/wiki/How-to-Run-a-Release) on How to Run a Release?

File: .gitignore (hunk 1 above)
Comment (hokei) [reply .gitignore:7]: Is the issue that https://github.com/NYPL/nypl-design-system/pull/71 resolves no longer relevant?

File: .github/workflows/release.yml (hunk 2 above)
Comment (hokei) [reply .github/workflows/release.yml:7]: Yep that's exactly what it is!

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+              with:
+                  tag_name: ${{ steps.autotagger.outputs.tagname }}
+                  release_name: ${{ steps.autotagger.outputs.tagname }}
+                  # use the body of the PR commit as the release body
+                  body: $commitmsg
+                  draft: false
+                  prerelease: false
+
+    publish_to_npm:
```
Comment (hokei) [inline .github/workflows/release.yml:47]: Do you know if it will need permissions to publish inside the @nypl org?

File: .github/workflows/release.yml (hunk 2 above)
Comment (helenvholmes) [reply .github/workflows/release.yml:7]: Do these branches exist in perpetuity in addition to the tagged releases?

File: .gitignore (hunk 1 above)
Comment (helenvholmes) [reply .gitignore:7]: Looks so because it'll be running those two other scripts is my guess

**Comment by kristojorg** [comment]
Yes I will! I also am not entirely sure this is bug free yet. Sadly I don’t know how to test it without merging and running the workflow

**Comment by hokei** [comment]
yeah, merge and tweak -- I think that's just how this goes!

**Review by hokei — APPROVED** [review]

**Review by hokei — CHANGES_REQUESTED** [review]
sorry, missed Helen's request for changes above

File: .github/workflows/release.yml (hunk 2 above)
Comment (kristojorg) [reply .github/workflows/release.yml:7]: Oh actually I think this is a little different. We won't have a `master` branch, but instead have a single `release` branch, which is the currently released code. The difference between calling it `master` or `release` is just semantic. There _won't_ be a separate branch for every release (like `release-v0.6.5`), which is what I think you were referring to. The purpose of marking/saving the state of the code when a release was made is what git tag is for. So if we ever needed to go back and branch off from an old release, we would simply start from the tag for that release. 

Long story short, we'll have:
- `release`
- `development` (default)
- `feature-branches...`

And a tag for every version:
- `v0.6.3`
- `v0.6.4`
- etc

File: .github/workflows/release.yml (hunk 3 above)
Comment (kristojorg) [reply .github/workflows/release.yml:47]: Yep, we need to add a secret to the repo called `NPM_TOKEN`

File: .gitignore (hunk 1 above)
Comment (kristojorg) [reply .gitignore:7]: Yea I just read through that PR. It _should_ now be possible to gitignore this folder, but I will have to double check once we merge this that it worked (I think it will)

**Comment by kristojorg** [comment]
Okay I made an update to the wiki page. Let me know what you think!

File: .github/workflows/release.yml (hunk 2 above)
Comment (helenvholmes) [reply .github/workflows/release.yml:7]: Ahh, that makes sense. Thanks for the explanation Kristo!

**Review by helenvholmes — APPROVED** [review]
Since we're not sure about whether or not we can gitignore .storybook-static I think it's fine to leave it in the README for now and follow-up after we've tested the workflow. Thanks for this hard work Kristo!

**Comment by kristojorg** [comment]
@hokei I think I still need you to approve before I can merge since you also requested changes : )

**Comment by helenvholmes** [comment]
Looks like the new checks are running on other PRs, so let's try and get this merged and ironed out to not delay parallel work :)

**Review by hokei — APPROVED** [review]
